File: cpp/fury/util/benchmark.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <benchmark/benchmark.h>

#include <codecvt>
#include <locale>
#include <random>

#include "string_util.h"

#include <cstring>
#include <string>

// Function to generate a random UTF-16 string
std::u16string generateRandomUTF16String(size_t length) {
  const char charset[] =
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
  std::default_random_engine rng(std::random_device{}());
  std::uniform_int_distribution<> dist(0, sizeof(charset) - 2);

  std::u16string result;
  result.reserve(length);
  for (size_t i = 0; i < length; ++i) {
    result += static_cast<char16_t>(charset[dist(rng)]);
  }

  return result;
}

std::vector<std::u16string> generateUTF16String(size_t num_tests) {
  std::vector<std::u16string> test_strings;
  for (size_t i = 0; i < num_tests; ++i) {
    test_strings.push_back(generateRandomUTF16String(num_tests));
  }
  return test_strings;
}

const std::vector<std::u16string> test_strings = generateUTF16String(1000);

// UTF16 to UTF8 using the standard library
std::string utf16ToUtf8StandardLibrary(const std::u16string &utf16) {
  std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t> convert;
  return convert.to_bytes(utf16);
}

// UTF16 to UTF8 baseline conversion (without SIMD)
std::string utf16ToUtf8BaseLine(const std::u16string &utf16,
                                bool is_little_endian = true) {
  size_t utf16_length = utf16.length();
  size_t utf8_length = utf16_length * 3;
  std::string utf8_result(utf8_length, '\0');

  size_t i = 0, j = 0;
  while (i < utf16_length) {
    char16_t utf16_char = utf16[i++];
    if (utf16_char < 0x80) {
      utf8_result[j++] = static_cast<char>(utf16_char);
    } else if (utf16_char < 0x800) {
      utf8_result[j++] = static_cast<char>(0xC0 | (utf16_char >> 6));
      utf8_result[j++] = static_cast<char>(0x80 | (utf16_char & 0x3F));
    } else {
      utf8_result[j++] = static_cast<char>(0xE0 | (utf16_char >> 12));
      utf8_result[j++] = static_cast<char>(0x80 | ((utf16_char >> 6) & 0x3F));
      utf8_result[j++] = static_cast<char>(0x80 | (utf16_char & 0x3F));
    }
  }

  utf8_result.resize(j);
  return utf8_result;
}

// Benchmark function for Standard Library UTF-16 to UTF-8 conversion
static void BM_StandardLibrary(benchmark::State &state) {
  for (auto _ : state) {
    for (const auto &str : test_strings) {
      std::string utf8 = utf16ToUtf8StandardLibrary(str);
    }
  }
}

// Benchmark function for Baseline UTF-16 to UTF-8 conversion
static void BM_BaseLine(benchmark::State &state) {
  for (auto _ : state) {
    for (const auto &str : test_strings) {
      std::string utf8 = utf16ToUtf8BaseLine(str, true);
    }
  }
}

// Benchmark function for SIMD-based UTF-16 to UTF-8 conversion
static void BM_SIMD(benchmark::State &state) {
  for (auto _ : state) {
    for (const auto &str : test_strings) {
      std::string utf8 = fury::utf16ToUtf8(str, true);
    }
  }
}

BENCHMARK(BM_StandardLibrary);
BENCHMARK(BM_BaseLine);
BENCHMARK(BM_SIMD);
BENCHMARK_MAIN();

File: integration_tests/graalvm_tests/src/main/java/org/apache/fory/graalvm/Benchmark.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fory.graalvm;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputFilter;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.apache.fory.Fory;
import org.apache.fory.io.ClassLoaderObjectInputStream;
import org.apache.fory.util.Preconditions;

/** Benchmark suite for fory and jdk. */
public class Benchmark {
  static ObjectInputFilter filter =
      ObjectInputFilter.Config.createFilter("org.apache.fory.graalvm.*");

  public static Object testJDKSerialization(Object o) {
    byte[] bytes = jdkSerialize(o);
    return jdkDeserialize(bytes);
  }

  public static byte[] jdkSerialize(Object data) {
    ByteArrayOutputStream bas = new ByteArrayOutputStream();
    jdkSerialize(bas, data);
    return bas.toByteArray();
  }

  public static void jdkSerialize(ByteArrayOutputStream bas, Object data) {
    bas.reset();
    try (ObjectOutputStream objectOutputStream = new ObjectOutputStream(bas)) {
      objectOutputStream.writeObject(data);
      objectOutputStream.flush();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static Object jdkDeserialize(byte[] data) {
    try (ByteArrayInputStream bis = new ByteArrayInputStream(data);
        ObjectInputStream objectInputStream =
            new ClassLoaderObjectInputStream(Thread.currentThread().getContextClassLoader(), bis)) {
      return objectInputStream.readObject();
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  private static final Fory fory1;
  private static final Fory fory2;

  static {
    fory1 = Fory.builder().withNumberCompressed(false).build();
    fory1.register(Foo.class, true);
    fory1.register(Struct.class, true);
    fory2 = Fory.builder().withNumberCompressed(true).build();
    fory2.register(Foo.class, true);
    fory2.register(Struct.class, true);
  }

  public static void main(String[] args) {
    List<String> list = new ArrayList<>();
    for (int i = 0; i < 20; i++) {
      list.add("string" + i);
    }
    Map<String, Long> map = new HashMap<>();
    for (int i = 0; i < 20; i++) {
      map.put("key" + i, (long) i);
    }
    benchmark(true, Struct.create());
    benchmark(false, Struct.create());
    benchmark(true, new Foo(100, "abc", list, map));
    benchmark(false, new Foo(100, "abc", list, map));
  }

  public static void benchmark(boolean compressNumber, Object obj) {
    String foryRepeat = System.getenv("BENCHMARK_REPEAT");
    if (foryRepeat == null) {
      return;
    }
    int n = Integer.parseInt(foryRepeat);
    Fory fory = compressNumber ? fory2 : fory1;
    System.out.println("=========================");
    System.out.println("Benchmark repeat number: " + foryRepeat);
    System.out.println("Object type: " + obj.getClass());
    System.out.println("Compress number: " + compressNumber);
    double forySize = fory.serialize(obj).length;
    System.out.println("Fory size: " + forySize);
    double jdkSize = jdkSerialize(obj).length;
    System.out.println("JDK size: " + jdkSize);
    Object o = fory.deserialize(fory.serialize(obj));
    for (int i = 0; i < n; i++) {
      o = fory.deserialize(fory.serialize(obj));
      testJDKSerialization(o);
    }
    long start = System.nanoTime();
    for (int i = 0; i < n; i++) {
      o = fory.deserialize(fory.serialize(o));
    }
    long duration = (System.nanoTime() - start);
    System.out.println("Fory serialization took mills: " + duration / 1000_000);
    Preconditions.checkArgument(o.equals(obj));

    start = System.nanoTime();
    for (int i = 0; i < n; i++) {
      o = testJDKSerialization(o);
    }
    long duration2 = (System.nanoTime() - start);
    System.out.println("JDK serialization took mills: " + duration2 / 1000_000);
    Preconditions.checkArgument(o.equals(obj));
    System.out.printf("Compare speed: Fory is %.2fx speed of JDK\n", duration2 * 1.0d / duration);
    System.out.printf("Compare size: Fory is %.2fx size of JDK\n", forySize / jdkSize);
  }
}
